libevmasm/Assembly.cpp

```diff
@@ -679,13 +679,30 @@ LinkerObject const& Assembly::assemble() const
 		// Append an INVALID here to help tests find miscompilation.
 		ret.bytecode.push_back(static_cast<uint8_t>(Instruction::INVALID));
 
+	map<h256, size_t> subAssemblyOffsets;
+	map<h256, LinkerObject> subAssemblyLinkerObjects;
```

Comment:
Maybe it's easier to just make this a ``std::map<LinkerObject, size_t> linkerObjectOffsets;`` and
define a ``<``comparison on ``LinkerObject``? That should be rather straightforward and immediately only give you full matches.
We could also start hashing full linker objects, but comparing is probably easier.

Reply:
The current implementation is off if you have three assemblies A, B and C, they all have identical bytecode, you end up processing A first, storing its linker object with the bytecode hash, but B and C have the same linker&immutable references, but different ones from A. Then you could deduplicate B and C, but you won't, since you chose A as reference...
Maybe easiest to avoid such kinds of corner-case non-sense by just not manually hashing at all and just using a binary true for comparing linker objects themselves - should be fast enough.

Reply:
Exactly, but thinking about it, I started questioning why we were hashing the bytecode for comparisons in the first place, because technically, `linkerObjectA.bytecode == linkerObjectB.bytecode` should work just as well, no? In any case, I was too embarrassed to asked, so just assumed that there is in fact a good reason for the hashing check.

Reply:
Technically, going via hashes is an optimization. And I put the hash version in the other PR before I started to think about the link references :-). But yeah, generally, if you didn't store this in some sort of container you need to compare each subassembly with all others really, which may be a bit too much... but storing in a binary tree is good enough as well.